lock.go

```diff
@@ -0,0 +1,716 @@
+	// new TTL for the locks will be the value of the argument provided.
+	//
+	// A LockStatus struct is returned for each lock that is renewed. The
+	// TTL field in each struct represents the new TTL. A client can use
+	// this information to ensure that all of the locks they created with
+	// the lockId have been renewed. If not all locks have been renewed,
+	// or if an error is returned, clients should assume that none of the
+	// locks are safe to use, and they should unlock the lockId.
+	Renew(lockId string, ttl uint) ([]LockStatus, error)
+}
+
+// mongoLocker implements the Locker interface with MongoDB.
+type mongoLocker struct {
+	coll *mgo.Collection
+}
+
+// NewMongoLocker creates a Locker that is backed by MongoDB.
+func NewMongoLocker(collection *mgo.Collection) Locker {
+	return &mongoLocker{
+		coll: collection,
+	}
+}
+
+func (l *mongoLocker) XLock(resourceName, lockId string, ld LockDetails) error {
+	selector := bson.M{
+		"resource":           resourceName,
+		"exclusive.acquired": false,
+		"shared.count":       0,
+	}
+
+	r := &resource{
+		Name:      resourceName,
+		Exclusive: lockFromDetails(lockId, ld),
+		Shared: sharedLocks{
+			Count: 0,
+			Locks: []lock{},
+		},
+	}
+
+	change := mgo.Change{
+		Update: r,
+		Upsert: true,
+	}
+
+	// One of three things will happen when we run this change (upsert).
+	// 1) If the resource exists and has any locks on it (shared or
+	//    exclusive), we will get a duplicate key error.
+	// 2) If the resource exists but doesn't have any locks on it, we will
+	//    update it to obtain an exclusive lock.
+	// 3) If the resource doesn't exist yet, it will be inserted which will
+	//    give us an exclusive lock on it.
+	_, err := l.coll.Find(selector).Apply(change, map[string]interface{}{})
+	if err != nil {
+		if mgo.IsDup(err) {
+			return ErrAlreadyLocked
+		}
+		return err
+	}
+
+	// Acquired lock.
+	return nil
+}
+
+func (l *mongoLocker) SLock(resourceName, lockId string, ld LockDetails, maxConcurrent int) error {
+	selector := bson.M{
+		"resource":           resourceName,
+		"exclusive.acquired": false,
+		"shared.locks.lockId": bson.M{
+			"$ne": lockId,
+		},
+	}
+	if maxConcurrent >= 0 {
+		selector["shared.count"] = bson.M{
+			"$lt": maxConcurrent,
+		}
+	}
+
+	change := mgo.Change{
+		Update: bson.M{
+			"$inc": bson.M{
+				"shared.count": 1,
+			},
+			"$push": bson.M{
+				"shared.locks": lockFromDetails(lockId, ld),
+			},
+		},
+		Upsert: true,
+	}
+
+	// One of three things will happen when we run this change (upsert).
+	// 1) If the resource exists and has an exclusive lock on it, or if it
+	//    exists and has a shared lock on it with the same lockId, we will
+	//    get a duplicate key error.
+	// 2) If the resource exists but doesn't have an exclusive lock or a
+	//    a shared lock with the same lockId on it, we will update it to
+	//    obtain a shared lock so long as there are less than maxConcurrent
+	//    shared locks on it already.
```

Comment:
Let's say a process first gets a SLock with maxConcurrent of 2. Later processes could then acquire an SLock on the same resource by giving a maxConcurrent value > 2. What guarantees the original request for a max number of concurrent locks on the resource?

Reply:
It is the responsibility of the caller to ensure that they are picking a suitable value for maxConcurrent. There is nothing in this package to guarantee the maxConcurrent value in the original request will be respected in the future, and that's on purpose.

A nice thing about doing it this way is that maxConcurrent isn't set in stone for any given shared lock. Say you created a shared lock on a resource and set maxConcurrent to 1, but you later decide that actually 3 shared locks can co-exist on it. In this case you don't have to wait around for the existing lock to finish and then modify the document to say "allow up to 3 shared locks" - all you have to do is just start using 3 for maxConcurrent in subsequent lock requests.

Reply:
Ok. I think the statement in the Readme "It is possible to limit the number of shared locks that can be on a resource at a time." can be made clearer then.

Reply:
ack, will update the readme. I'm still trying to figure out how much info I want to include there vs. just having people read the godocs, and perhaps now I'm at some middle state where the readme has enough info to make it seem like the only thing you need to read, but in reality not enough info to properly use the package.

Reply:
Updated the readme to point to the godocs.